python/observation-publisher/publisher/prediction_log_consumer.py

```diff
@@ -0,0 +1,150 @@
+import abc
+from typing import List, Optional, Union, Dict
+
+import numpy as np
+import pandas as pd
+from caraml.upi.v1.prediction_log_pb2 import PredictionLog
+from confluent_kafka import Consumer, KafkaException
+from google.protobuf.internal.well_known_types import ListValue
+
+from publisher.config import (
+    KafkaConsumerConfig,
+    ModelSchema,
+    ModelSpec,
+    ValueType,
+    PredictionLogConsumerConfig,
+    ConsumerType,
+)
+from publisher.observability_backend import ObservationSink
+
+
+class PredictionLogConsumer(abc.ABC):
+    @abc.abstractmethod
+    def poll_new_logs(self) -> List[PredictionLog]:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def commit(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def close(self):
+        raise NotImplementedError
+
+    def start_polling(self, observation_sink: ObservationSink, model_spec: ModelSpec):
+        try:
+            while True:
+                logs = self.poll_new_logs()
+                if len(logs) == 0:
+                    continue
+                df = log_to_dataframe(logs, model_spec.schema)
+                observation_sink.write(dataframe=df)
+                self.commit()
+        finally:
+            self.close()
+
+
+class KafkaPredictionLogConsumer(PredictionLogConsumer):
+    def __init__(self, config: KafkaConsumerConfig):
+        consumer_config = {
+            "bootstrap.servers": config.bootstrap_servers,
+            "group.id": config.group_id,
+        }
+
+        if config.additional_consumer_config is not None:
+            consumer_config.update(config.additional_consumer_config)
+
+        self._consumer = Consumer(consumer_config)
+        self._batch_size = config.batch_size
+        self._consumer.subscribe([config.topic])
+        self._poll_timeout = config.poll_timeout_seconds
+
+    def poll_new_logs(self) -> List[PredictionLog]:
+        messages = self._consumer.consume(self._batch_size, timeout=self._poll_timeout)
+        errors = [msg.error() for msg in messages if msg.error() is not None]
+
+        if len(errors) > 0:
+            print(f"Last encountered error: {errors[-1]}")
+            raise KafkaException(errors[-1])
+
+        return [
+            parse_message_to_prediction_log(msg.value().decode("utf-8"))
+            for msg in messages
+            if (msg is not None and msg.error() is None)
+        ]
+
+    def commit(self):
+        self._consumer.commit()
+
+    def close(self):
+        self._consumer.close()
+
+
+def new_consumer(config: PredictionLogConsumerConfig) -> PredictionLogConsumer:
+    if config.type == ConsumerType.KAFKA:
+        return KafkaPredictionLogConsumer(config.kafka_config)
+    else:
+        raise ValueError(f"Unknown consumer type: {config.type}")
+
+
+def parse_message_to_prediction_log(msg: str) -> PredictionLog:
+    log = PredictionLog()
+    log.ParseFromString(msg)
+    return log
+
+
+def convert_value(
```

Comment:
Nit: Would it be clearer to have `convert_value_to_np` or something like that to indicate what conversion is happening? Same comment about the list conversion function below.